internal/bcdb/transaction_processor.go

```diff
@@ -266,37 +265,36 @@ func (t *transactionProcessor) SubmitTransaction(tx interface{}, timeout time.Du
 		return nil, err
 	}
 
-	t.Lock()
-	duplicate, err := t.isTxIDDuplicate(txID)
-	if err != nil {
-		t.Unlock()
-		return nil, err
-	}
-	if duplicate {
-		t.Unlock()
+	// We attempt to insert the txID atomically.
+	// If we succeed, then future TX fill fail at this point.
+	// However, if the TX already exists in the block store, then we will fail the subsequent check.
+	// Since a TX will be removed from the pending queue only after it is inserted to the block store,
+	// Then it is guaranteed that we won't use the same txID twice.
+	// TODO: add limit on the number of pending sync tx
+	promise := queue.NewCompletionPromise(timeout)
+	if existed := t.pendingTxs.Add(txID, promise); existed {
```

Comment:
Why did you remove this part of the code?

```
	if t.pendingTxs.Has(txID) {
 		return true, nil
 	}
```

We want to ensure that there is no duplicate txID either in block store or provenance store. Hence, we reject any transaction that has a duplicate txID with either pendingTxs or block store. At all time, our block store has unique txIDs. Whoever we accept a transaction, we check whether the txID is duplicate by looking at all pending txs and committed states. Only if it is found to be unique, we would accept the transaction. Otherwise, it is rejected.

Reply:
`t.pendingTxs.Add(txID, promise)` will return `existed=true` if it was already in the map.
So this one atomic call replaces the two non-atomic calls, avoiding the [TocTou](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue.

Reply:
got it. thanks.